Stratis.Bitcoin/BlockPulling/LookaheadBlockPuller.cs

```diff
@@ -385,38 +428,46 @@ private Block NextBlockCore(CancellationToken cancellationToken)
                 {
                     if (header.Previous.HashBlock != this.location.HashBlock)
                     {
-                        //A reorg
-                        return null;
+                        this.logger.LogTrace("Blockchain reorganization detected.");
+                        break;
                     }
                     this.IsStalling = false;
                     this.location = header;
                     Interlocked.Add(ref this.currentSize, -block.Length);
                     this.consumed.Set();
-                    return block.Block;
+
+                    res = block.Block;
                 }
                 else
                 {
-                    // Otherwise we either have reorg.
+                    // Otherwise we either have reorg, or we don't know the next block.
                     if (header == null)
                     {
                         if (!this.Chain.Contains(this.location.HashBlock))
                         {
-                            //A reorg
-                            return null;
+                            this.logger.LogTrace("Blockchain reorganization detected.");
+                            break;
                         }
+
+                        this.logger.LogTrace("Hash of the next block is not known.");
                     }
                     else
                     {
+                        this.logger.LogTrace($"Block not available.");
+
                         // Or the block is still being downloaded or we need to ask for this block to be downloaded.
                         if (!isDownloading) AskBlocks(new ChainedBlock[] { header });
 
                         OnStalling(header);
                         this.IsStalling = true;
                     }
-                    WaitHandle.WaitAny(new[] { this.pushed, cancellationToken.WaitHandle }, waitTime[i]);
-                    i = i == waitTime.Length - 1 ? 0 : i + 1;
+
+                    WaitHandle.WaitAny(new[] { this.pushed, cancellationToken.WaitHandle }, WaitNextBlockRoundTimeMs);
```

Comment:
Ok yeah true once not in IBD the node puller wont be so busy (but mempool might be hammered very hard)

But I don't see any complexity, just a simple flag on the if statement above.

```
WaitHandle.WaitAny(new[] { this.pushed, cancellationToken.WaitHandle }, waitIBD ?
 WaitNextBlockRoundIBDTimeMs : WaitNextBlockRoundTimeMs );
```

I am not going to insist thought because I am not sure myself if it will effect propagation.